File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
// TC - N
// SC - N 
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        queue<TreeNode*> q;
        int lvl = 0; 

        if (root == NULL) {
            return res;
        }
        q.push(root);

        while (!q.empty()) {
            vector<int> vis;
            int size = q.size();
            lvl++;   //  not inside for loop  // 
            for (int i = 0; i < size; i++) {
                TreeNode* temp = q.front();
                q.pop();
                vis.push_back(temp->val);

                if (temp->left) {
                    q.push(temp->left);
                }
                if (temp->right) {
                    q.push(temp->right);
                }
            }
            if (lvl % 2 == 0) {
                reverse(vis.begin(), vis.end());
            }
            res.push_back(vis);

        }
        return res;
    }

};

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        int level = 0;

        if(root == null) {
            return res;
        }
        q.offer(root);

        while(!q.isEmpty()) {
            level++;
            int size = q.size();
            List<Integer> temp = new ArrayList<>();
            for(int i = 0; i <  size; i++) {
                TreeNode node = q.poll();
                temp.add(node.val);
                if(node.left != null) {
                    q.add(node.left);
                }
                if(node.right != null ) {
                    q.add(node.right);
                }
            }
            if(level%2 == 0) {
                Collections.reverse(temp);  
            }
            res.add(temp);
        }
        return res;
    }
}